modules/dnn/src/onnx/onnx_graph_simplifier.cpp

```diff
@@ -290,9 +290,112 @@ class ResizeSubgraph2 : public ExtractScalesSubgraph
     }
 };
 
+class PaddingSubgraph : public Subgraph
+{
+public:
+    PaddingSubgraph()
+    {
+        input = addNodeToMatch("");
+        constantPadding = addNodeToMatch("Constant");
+        int constant2 = addNodeToMatch("Constant");
+        int shape = addNodeToMatch("Shape", constantPadding);
+        int gather = addNodeToMatch("Gather", shape, constant2);
+        int constantMulA = addNodeToMatch("Constant");
+        int constantMulB = addNodeToMatch("Constant");
+        int mul = addNodeToMatch("Mul", constantMulA, constantMulB);
+        int sub = addNodeToMatch("Sub", mul, gather);
+        int cast = addNodeToMatch("Cast", constantPadding);
+        int constantofshape = addNodeToMatch("ConstantOfShape", sub);
+        int concat = addNodeToMatch("Concat", cast, constantofshape);
+        int constantRC = addNodeToMatch("Constant");
+        int reshapeConcat = addNodeToMatch("Reshape", concat, constantRC);
+        int constantAxes = addNodeToMatch("Constant");
+        int constantStart = addNodeToMatch("Constant");
+        int constantEnds = addNodeToMatch("Constant");
+        int constantSteps = addNodeToMatch("Constant");
+
+        int sliceInputs[] = {reshapeConcat, constantStart, constantEnds, constantAxes, constantSteps};
+
+        int slice = addNodeToMatch("Slice", std::vector<int>(&sliceInputs[0], &sliceInputs[0] + 5));
+        int transpose = addNodeToMatch("Transpose", slice);
+        int constantRT = addNodeToMatch("Constant");
+        int reshapeT = addNodeToMatch("Reshape", transpose, constantRT);
+        castRT = addNodeToMatch("Cast", reshapeT);
```

Comment:
Can onnx_importer computer this tensor?

Reply:
@dkurt No, it cannot. 

I tried some experiments, made some changes but ended up with this error at slice layer. 

```
cv2.error: OpenCV(3.4.10-dev) /home/ashish/opencv/modules/dnn/src/onnx/onnx_importer.cpp:485: error: (-2:Unspecified error) in function 'void cv::dnn::experimental_dnn_34_v17::ONNXImporter::populateNet(cv::dnn::experimental_dnn_34_v17::Net)'
> Slice layer only supports steps = 1 (expected: 'countNonZero(step_blob != 1) == 0'), where
>     'countNonZero(step_blob != 1)' is 1
> must be equal to
>     '0' is 0
```
Because slice layer has inputs shown in the image.
![Screenshot from 2020-04-06 10-56-45](https://user-images.githubusercontent.com/26292819/78525653-67c6dc00-77f5-11ea-9b34-fde16983796a.png)

**Change I made**
Change 1. allow shape to take inpshape from constBlob
```cpp
else if (layer_type == "Shape")
        {
            CV_Assert(node_proto.input_size() == 1);
            shapeIt = outShapes.find(node_proto.input(0));

            MatShape inpShape = shapeIt != outShapes.end() ? shapeIt->second : shape(getBlob(node_proto, constBlobs, 0));
            Mat shapeMat(inpShape.size(), 1, CV_32S);
            for (int j = 0; j < inpShape.size(); ++j)
                shapeMat.at<int>(j) = inpShape[j];
            shapeMat.dims = 1;

            constBlobs.insert(std::make_pair(layerParams.name, shapeMat));
            continue;
        }
```
from 
```cpp
else if (layer_type == "Shape")
        {
            CV_Assert(node_proto.input_size() == 1);
            shapeIt = outShapes.find(node_proto.input(0));
            CV_Assert(shapeIt != outShapes.end());
            MatShape inpShape = shapeIt->second;

            Mat shapeMat(inpShape.size(), 1, CV_32S);
            for (int j = 0; j < inpShape.size(); ++j)
                shapeMat.at<int>(j) = inpShape[j];
            shapeMat.dims = 1;

            constBlobs.insert(std::make_pair(layerParams.name, shapeMat));
            continue;
        }
```


Change 2. make all input datatypes same. 
input 1 from cast was int and input 2 from constantofshape was float32, i guess, so was giving error. 
```cpp
else if (layer_type == "Concat")
        {
            bool hasVariableInps = false;
            for (int i = 0; i < node_proto.input_size(); ++i)
            {
                if (layer_id.find(node_proto.input(i)) != layer_id.end())
                {
                    hasVariableInps = true;
                    break;
                }
            }

            if (!hasVariableInps)
            {
                std::vector<Mat> inputs(node_proto.input_size()), concatenated;
                for (size_t i = 0; i < inputs.size(); ++i)
                {
                    inputs[i] = getBlob(node_proto, constBlobs, i);
                    inputs[i].convertTo(inputs[i], CV_32F);
                }
                runLayer(layerParams, inputs, concatenated);

                CV_Assert(concatenated.size() == 1);
                constBlobs.insert(std::make_pair(layerParams.name, concatenated[0]));
                continue;
            }
        }
```
from 
```cpp
else if (layer_type == "Concat")
        {
            bool hasVariableInps = false;
            for (int i = 0; i < node_proto.input_size(); ++i)
            {
                if (layer_id.find(node_proto.input(i)) != layer_id.end())
                {
                    hasVariableInps = true;
                    break;
                }
            }

            if (!hasVariableInps)
            {
                std::vector<Mat> inputs(node_proto.input_size()), concatenated;
                for (size_t i = 0; i < inputs.size(); ++i)
                {
                    inputs[i] = getBlob(node_proto, constBlobs, i);
                }
                runLayer(layerParams, inputs, concatenated);

                CV_Assert(concatenated.size() == 1);
                constBlobs.insert(std::make_pair(layerParams.name, concatenated[0]));
                continue;
            }
        }
```


change 3. if data is less than min or greater than max, dont give error. 

```cpp
void convertInt64ToInt32(const T1& src, T2& dst, int size)
{
    for (int i = 0; i < size; i++)
    {
        if (src[i] < std::numeric_limits<int32_t>::min())
        {
            dst[i] = saturate_cast<int32_t>(std::numeric_limits<int32_t>::min());
            continue;
        }
        if (src[i] > std::numeric_limits<int32_t>::max())
        {
            dst[i] = saturate_cast<int32_t>(std::numeric_limits<int32_t>::max());
            continue;
        }
        dst[i] = saturate_cast<int32_t>(src[i]);
    }
}
```
from 
```cpp
void convertInt64ToInt32(const T1& src, T2& dst, int size)
{
    for (int i = 0; i < size; i++) {
        if (src[i] < std::numeric_limits<int32_t>::min() || src[i] > std::numeric_limits<int32_t>::max()) {
            CV_Error(Error::StsOutOfRange, "Input is out of OpenCV 32S range");
        }
        dst[i] = saturate_cast<int32_t>(src[i]);
    }
}
```

And after all that slice node gave above mentioned error.